src/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs

```diff
@@ -134,11 +134,22 @@ public bool ObserveCancellation()
 
             _awaitableState &= ~(AwaitableState.Canceled | AwaitableState.Running);
 
-            _cancellationToken.ThrowIfCancellationRequested();
-
             return isCanceled;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public CancellationTokenRegistration ReleaseCancellationTokenRegistration()
+        {
+            CancellationToken cancellationToken = _cancellationToken;
+            CancellationTokenRegistration cancellationTokenRegistration = _cancellationTokenRegistration;
+            _cancellationTokenRegistration = default;
+            _cancellationToken = default;
+
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
I get that the registered callback was likely called if cancellation was requested, but what if it wasn't? Couldn't there be a race where Reader/WriterCancellationRequested() gets called at some arbitrary point in the future, maybe even after a subsequent read has started, since the registration isn't disposed in this case?